[Ide][Core] Adds support for AppendTargetFrameworkToOutputPath

jtorres opened the pull request:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/810355

jtorres committed:
[Ide][Core] Adds support for AppendTargetFrameworkToOutputPath

mrward commented:
Tested a local build of this. Looking better. The problems I found testing were:

1) Opening project options - Compiler and then OK's the dialog without making changes the project file:

```
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <OutputPath>bin\Debug</OutputPath>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <OutputPath>bin\Release</OutputPath>
  </PropertyGroup>
```

Was not expecting any changes to the project file.

2) Editing Debug path so it uses bin2 then seems to affect the Release output path. The Release path no longer includes the target framework. I was expecting the Release path to still be using a target framework. Is that possible/supported?

3) You cannot use $(TargetFramework) anywhere in the path. Not sure if we want to support that or not. I am OK with not supporting this.

rodrmoya commented:
The test failures seem to indicate indeed that the csproj is being changed?

jtorres committed:
[Ide] Always checks AppendTargetFrameworkToOutPath state

jtorres commented:
@mrward :

1. Got it, fix in the next commit
2. Can't reproduce it... in any case, if you remove the target framework `AppendTargetFrameworkToOutputPath = false` should be added
3. Yes  you can use `$(TargetFramework)` in any configuration

mrward commented:
2) Will re-test a build from this branch. I did not remove the $(TargetFramework). Just changed bin to bin2 for the Debug configuration only.

3) If I copy $(TargetFramework) just after /bin/ then it is removed.

    /full/path/to/bin/$(TargetFramework)/$(Configuration)/$(TargetFramework)

That seems to lose the $(TargetFramework) after the bin/ directory.

Not sure if that is supported.

jtorres commented:
3. retest it, this is what I am getting: 
![Screenshot 2019-07-02 10 32 25](https://user-images.githubusercontent.com/283707/60521206-d00bee80-9cb4-11e9-8367-48907ce03af6.png)

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProjectConfiguration.cs:
```diff
@@ -79,6 +85,11 @@ internal protected override void Write (IPropertySet pset)
+			if (AppendTargetFrameworkToOutputPath) {
+				pset.RemoveProperty (nameof (AppendTargetFrameworkToOutputPath));
+			} else {
+				pset.SetValue (nameof (AppendTargetFrameworkToOutputPath), false);
```
I suspect this is what is breaking the unit tests. We should not be adding AppendTargetFrameworkToOutputPath if it did not exist before in the set of known properties.

Maybe make AppendTargetFrameworkToOutputPath a `bool?` and check if it has a value before trying to set it to false.

jtorres replied:
Good point,, just fixed, all green 👌

jtorres committed:
[Ide] Turns AppendTargetFrameworkToOutput prop nullable

So that only when .NET Core / Standard project is loaded a boolean value is set. Otherwise is null, that's .NET Fx

rodrmoya commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProjectConfiguration.cs:
```diff
@@ -81,55 +78,41 @@ internal protected override void Write (IPropertySet pset)
+			if (compilationParameters != null)
+				compilationParameters.Write (pset);
+
+			if (AppendTargetFrameworkToOutputPath == null || AppendTargetFrameworkToOutputPath == true) {
```
this can be shortened into `if (AppendTargetFrameworkToOutputPath ?? true)`

jtorres commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Projects.OptionPanels/OutputOptionsPanel.cs:
```diff
@@ -180,13 +145,140 @@ public void Store ()
+				// but TargetFrameworkId values, therefore we have to apply the same logic checking the end of the dir template
+				if (dir.TrimEnd (IO.Path.DirectorySeparatorChar).EndsWith (conf.TargetFrameworkShortName, StringComparison.InvariantCulture)) {
+					dir = dir.Replace (conf.TargetFrameworkShortName, string.Empty);
+				}
```
this is actually the key, what actually fixes the bug. $(TargetFramework) is only visible (whether that variable or the value itself) in the OptionPanel and msbuild would add or not depending on the AppendTargetFrameworkToOutputPath value.

jtorres commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Projects.OptionPanels/OutputOptionsPanel.cs:
```diff
@@ -180,13 +145,140 @@ public void Store ()
+
+			// if outputDirectory does not contain the targetFramework.Id, AppendTargetFrameworkToOutputPath is false for that config
+			conf.AppendTargetFrameworkToOutputPath = dir.Contains (conf.TargetFrameworkShortName) || dir.Contains ("$(TargetFramework)");
+
```
yep, will check only at the end...

jtorres commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Projects.OptionPanels/OutputOptionsPanel.cs:
```diff
@@ -180,13 +145,140 @@ public void Store ()
+			if (outputModified) { 
+				// if so, we have to remove $(TargetFramework) since msbuild will add it due to AppendTargetFrameworkToOutputPath == true 
+				dir = dir.Replace ("$(TargetFramework)", string.Empty);
+				// in case we are in a specific configuration i.e. Debug, Release, there will be no msbuild variable $(TargetFramework)
```
ditto (yep, will check only at the end...)

jtorres commented on main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Projects.OptionPanels/OutputOptionsPanel.cs:
```diff
@@ -180,13 +145,140 @@ public void Store ()
+				//the output part due to AppendTargetFrameworkToOutputPath == true
+				if (!dirTemplate.Contains (conf.TargetFrameworkShortName)) {
+					dirTemplate = IO.Path.Combine (dirTemplate, conf.TargetFrameworkShortName);
+				}
```
ditto (yep, will check only at the end...)

mrward commented:
Testing the latest build.

1) $(TargetFramework) cannot be added into the middle of the path - it gets removed. Mentioned this before.

    /Users/matt/Projects/Tests/MyProject/MyProject/bin/$(TargetFramework)/$(Configuration)/$(TargetFramework)

2) Prefixing $(Configuration) has problems.

'All configurations' changed to:

    /Users/matt/Projects/Tests/MyProject/MyProject/bin/test-$(Configuration)/$(TargetFramework)

Click OK.
Project file has the following which is OK.

```
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <OutputPath>bin\test-Debug</OutputPath>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <OutputPath>bin\test-Release</OutputPath>
  </PropertyGroup>
```

Re-open project options:

 All Configurations shows no output path.

Debug shows:

    /Users/matt/Projects/Tests/MyProject/MyProject/bin/test-Debug/netcoreapp2.1


I am assuming this is because we are looking for Path.Separator/$(Configuration) which seems to break the old behaviour in VS Mac.

jtorres committed:
[Ide] Few fixes:

- Method renaming
- Checks OutputDir is empty and fixes it.
- Uses Removes instead of Replace for EndWith strings

mrward commented:
Not seeing any code changes since the last time I reviewed/tested this. Did you push the changes?

jtorres commented:
@mrward pushed...

mrward commented:
Needs rebasing to fix the build failures.

jtorres commented:
@mrward I did rebase a couple of hours ago. It-s failing because of a random time out on dotnetcore templates tests...

mrward commented:
1) Change All Configurations so path is something like:

/Users/user/Projects/Tests/myproject/myproject/bin/$(TargetFramework)/$(Configuration)/$(TargetFramework)

Click OK. Expecting OutputPath to be bin\netcoreapp2.1\Debug and bin\netcoreapp2.1\Release

  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <OutputPath>bin\%24%28TargetFramework%29\Debug</OutputPath>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <OutputPath>bin\%24%28TargetFramework%29\Release</OutputPath>
  </PropertyGroup>

2) Path being reset.

Select Debug.
Change bin to bin2
Select All Configurations
Select Debug - bin2 changed back to bin. Path seems to be reset.

3) Prefix-configuration being reset.

All Configurations
Change $(Configuration) to foo-$(Configuration)
Switch to Debug
Switch to All Configurations - path is now blank.
Switch to Debug - path has been reset.

4) Prefixing $(Configuration) has problems.

'All configurations' changed to:

/Users/username/Projects/Tests/MyProject/MyProject/bin/test-$(Configuration)/$(TargetFramework)
Click OK.
Project file has the following which is OK.

  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <OutputPath>bin\test-Debug</OutputPath>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <OutputPath>bin\test-Release</OutputPath>
  </PropertyGroup>
Re-open project options:

All Configurations shows no output path.

jtorres committed:
[Ide] Gets rid of DirectorySeparator for replacing variables

jtorres commented:
Fixed

mrward commented:
Re-testing and most problems are fixed. Only problem I can still see is the path being reset.

Select Debug.
Change bin to bin2
Select All Configurations
Select Debug
bin2 changed back to bin. Path seems to be reset.

jtorres committed:
[Ide] Resets output to default only when there is no outputPathEntry in a given configuration
